File: tests/test_report_json.py
import json
import unittest

from painradar.models import Post
from painradar.clustering import Cluster
from painradar.opportunity import build_opportunities
from painradar.report_json import build_report_dict


def make_post(id_, author):
    return Post(
        id=id_,
        title=f"Manually doing invoice work {id_}",
        selftext="I manually copy invoice data every week, so frustrating, expensive, I'd pay for a fix.",
        subreddit="freelance",
        author=author,
        score=10,
        num_comments=5,
        created_utc=1,
        url=f"https://reddit.com/{id_}",
    )


class TestBuildReportDict(unittest.TestCase):
    def test_report_dict_is_json_serializable_and_has_expected_shape(self):
        cluster = Cluster(posts=[make_post("1", "a"), make_post("2", "b")])
        opportunities = build_opportunities([cluster])
        report = build_report_dict(opportunities, source_description="fixture: fixtures/sample_posts.json")

        serialized = json.dumps(report)
        parsed = json.loads(serialized)

        self.assertIn("generated_at", parsed)
        self.assertIn("source_description", parsed)
        self.assertIn("disclaimer", parsed)
        self.assertEqual(len(parsed["opportunities"]), 1)
        opp = parsed["opportunities"][0]
        self.assertIn("problem_statement", opp)
        self.assertIn("score", opp)
        self.assertIn("total", opp["score"])
        self.assertIn("reasons", opp["score"])
        self.assertIn("evidence", opp)
        self.assertIn("market_size_hypotheses", opp)


if __name__ == "__main__":
    unittest.main()

File: test/test_reportJson.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { Post } from "../src/models.js";
import { Cluster } from "../src/clustering.js";
import { buildOpportunities } from "../src/opportunity.js";
import { buildReportDict } from "../src/reportJson.js";

function makePost(id: string, author: string): Post {
  return {
    id,
    title: `Manually doing invoice work ${id}`,
    selftext: "I manually copy invoice data every week, so frustrating, expensive, I'd pay for a fix.",
    subreddit: "freelance",
    author,
    score: 10,
    num_comments: 5,
    created_utc: 1,
    url: `https://reddit.com/${id}`,
  };
}

test("report dict is JSON-serializable and has the expected shape", () => {
  const cluster = new Cluster([makePost("1", "a"), makePost("2", "b")]);
  const opportunities = buildOpportunities([cluster]);
  const report = buildReportDict(opportunities, "fixture: fixtures/sample_posts.json");

  const parsed = JSON.parse(JSON.stringify(report));

  assert.ok("generated_at" in parsed);
  assert.ok("source_description" in parsed);
  assert.ok("disclaimer" in parsed);
  assert.equal(parsed.opportunities.length, 1);
  const opp = parsed.opportunities[0];
  assert.ok("problem_statement" in opp);
  assert.ok("score" in opp);
  assert.ok("total" in opp.score);
  assert.ok("reasons" in opp.score);
  assert.ok("evidence" in opp);
  assert.ok("market_size_hypotheses" in opp);
});
